### PR TITLE
test(#910): turn-driver characterization tests (PR-1 of multi-PR refactor)

### PR DESCRIPTION
**Pull request by rysweet**
## Summary

PR-1 of the Issue #910 multi-PR refactor: a **characterization-only** safety net.
No production logic changes — every addition lives in existing test files or the
pre-existing `#[cfg(test)] mod tests` of `auto_mode_exec.rs` (excluded from
non-test builds). The suite locks the *current* observable behavior of the two
turn-driver loops so later refactor PRs can move code with a guardrail underneath.

Loops under characterization:
- **Signal in-process chat loop** — `crates/amplihack-cli/src/commands/signal/chat.rs`, driven by `SerialTurnDriver` + `CopilotTurnRunner` (`crates/amplihack-signal/src/chat/turn.rs`).
- **Auto-mode loop** — `crates/amplihack-launcher/src/auto_mode_exec.rs` (`AutoMode::run`).

All tests are hermetic: no real network/process, no `sleep`-as-synchronization.
Timing runs through explicit seams (`Gate::evaluate_at`, `record_outbound_at`),
ordering through channels/atomics.

## INV → test coverage (pre-existing vs added)

| INV | Invariant | Coverage | Test(s) | File |
| --- | --------- | -------- | ------- | ---- |
| INV-1 | Turn serialization | Pre-existing | `turn::turns_execute_one_at_a_time_per_session` | `crates/amplihack-signal/tests/chat_it.rs` |
| INV-2 | Post-on-completion only on verified membership | Pre-existing | `membership::{exact_expected_set_is_verified_and_may_relay, query_error_is_unverified_and_refuses_relay, unexpected_extra_member_refuses_relay, missing_expected_member_refuses_relay}`; `chat_membership_failclosed_it.rs` | `crates/amplihack-signal/tests/chat_it.rs`, `.../chat_membership_failclosed_it.rs` |
| INV-3 | Inbound gating fail-closed | Pre-existing | `gating::tests::{empty_allowlist_denies_everything, rejects_sender_not_on_allowlist, rejects_empty_body, rejects_message_for_other_group, rejects_non_group_message, rejects_non_primary_device_sync_echo, rejects_bot_own_linked_device_sync, rejects_sync_not_authored_by_account, suppresses_*_echo_within_ttl}` | `crates/amplihack-signal/src/gating.rs` |
| **INV-3** | Inbound gating fail-closed | **Added** | `characterization_inv3_inbound_gate_failclosed` (consolidated `fake_endpoint` anchor; deny-on-error/ambiguity, never fail-open, via `Gate::evaluate_at`) | `crates/amplihack-signal/tests/chat_it.rs` (`gating` mod) |
| INV-4 | Bounded turn queue evicts oldest at operator-configurable capacity | Pre-existing | `session_channel::tests::bounded_capacity_evicts_oldest`; `bounded_inbox_survives_a_flood` | `crates/amplihack-signal/src/session_channel.rs`, `.../tests/session_channel_it.rs` |
| **INV-4** | Bounded turn queue evicts oldest at operator-configurable capacity | **Added** | `characterization_inv4_capacity_from_operator_config_evicts_oldest` (`AMPLIHACK_SIGNAL_INBOX_CAPACITY` path: valid cap honored; invalid/zero/negative/non-numeric fall back to `DEFAULT_CAPACITY`; `PushOutcome::EvictedOldest` bounds memory) | `crates/amplihack-signal/tests/session_channel_it.rs` |
| **INV-5** | Resume continuity (same session id, injection-safe argv) | **Added** | `characterization_inv5_successive_turns_reuse_same_session_id` (same `--session-id`; prompt is exactly one `-p` argv element even with shell metacharacters) | `crates/amplihack-signal/tests/chat_it.rs` (`turn` mod) |
| **INV-6** | No per-turn wall-clock timeout | **Added** | `characterization_inv6_turn_runs_to_completion_no_wallclock_cap` (turn returns output on natural completion; no per-turn elapsed-time cap) | `crates/amplihack-signal/tests/chat_it.rs` (`turn` mod) |
| INV-7 | Auto-mode continues until stop condition | Pre-existing | `should_continue_respects_turns`, `should_continue_respects_api_calls` | `crates/amplihack-launcher/src/auto_mode_exec.rs` (`#[cfg(test)] mod tests`) |
| **INV-7** | Auto-mode continues until stop condition | **Added** | `characterization_inv7_automode_continues_until_stop_condition` (terminates on first tripped stop condition; `build_execution_prompt` emits 1-based `[Turn n/max]`) | `crates/amplihack-launcher/src/auto_mode_exec.rs` (`#[cfg(test)] mod tests`) |

> **INV-6 scope note:** INV-6 (no per-turn wall-clock timeout) is an intentional
> addition by this suite, not from the original spec. It is code-grounded
> (`SerialTurnDriver::run_turn` wraps the child in no `tokio::time::timeout`) and
> distinct from `max_session_secs` (a session-level cap, INV-7).

## Verification

- `cargo fmt --all --check` ✅
- `cargo clippy -p amplihack-signal --features signal --all-targets -- -D warnings` ✅
- `cargo clippy -p amplihack-launcher --all-targets -- -D warnings` ✅
- `cargo build -p amplihack-signal --features signal -p amplihack-launcher` ✅
- `cargo test -p amplihack-signal --features signal` ✅ (all 5 signal characterization tests + full suite pass)
- `cargo test -p amplihack-launcher characterization_inv7` ✅ (default build, no signal feature)

`git diff` shows only test files + `cfg(test)` helpers — no production logic change.

## Out of scope / notes

- Pre-existing, non-deterministic CLI test flakiness in unrelated files
  (multitask launcher CRLF, copilot_setup staging) is *not* touched by this PR and
  does not gate this characterization-only change.

Full write-up: `docs/testing/TURN_DRIVER_CHARACTERIZATION_TESTS.md`.

Closes part of #910.

## Step 13: Local Testing Results

Detected toolchains:
- **Rust / Cargo** (workspace). Evidence: root `Cargo.toml` workspace; changed files are all Rust test code — `crates/amplihack-signal/tests/{chat_it.rs,session_channel_it.rs}` (integration tests, `#![cfg(feature = "signal")]`) and a `#[cfg(test)] mod tests` block in `crates/amplihack-launcher/src/auto_mode_exec.rs`, plus a Markdown doc. The signal turn-driver is feature-gated (`--features signal`); auto-mode is in the default build.

Chosen validation strategy:
- Outside-in consumer boundary for a Rust CLI = the crate test suites exercising the public seams (`SerialTurnDriver`, injectable `TurnRunner`/`MockRunner`, `Gate::evaluate_at`, `Inbox` env path, `AutoMode` decision fns). Ran the five added `characterization_inv<N>_*` tests under the same feature flags as existing signal tests, plus the four cargo gates. This proves the observable turn-driver behavior is locked without touching production logic.

### Scenario 1 — simple: auto-mode INV-7 in the default build (no signal feature)
Command: `cargo test -p amplihack-launcher characterization_inv7_automode_continues_until_stop_condition`
Result: PASS
Output:
```
test auto_mode_exec::tests::characterization_inv7_automode_continues_until_stop_condition ... ok
test result: ok. 1 passed; 0 failed; 0 ignored; 0 measured; 299 filtered out
```

### Scenario 2 — complex: feature-gated signal turn-driver characterization suite (INV-3/4/5/6)
Command: `cargo test -p amplihack-signal --features signal characterization`
Result: PASS
Output:
```
gating::characterization_inv3_inbound_gate_failclosed ... ok
turn::characterization_inv5_successive_turns_reuse_same_session_id ... ok
turn::characterization_inv6_turn_runs_to_completion_no_wallclock_cap ... ok
characterization_inv4_capacity_from_operator_config_evicts_oldest ... ok
test result: ok. 3 passed (chat_it) + 1 passed (session_channel_it); 0 failed
```

### Cargo gates (all green)
- `cargo fmt --check` ✅
- `cargo clippy -p amplihack-signal -p amplihack-launcher --features signal --tests` ✅ (no warnings)
- `cargo build -p amplihack-signal -p amplihack-launcher --features signal` ✅
- `cargo test -p amplihack-signal --features signal` ✅ / `cargo test -p amplihack-cli --features signal signal_chat_it` ✅ (0 failed) / `cargo test -p amplihack-launcher` ✅

### Regression / no-production-change check
`git diff --name-only origin/main...HEAD` = only test files, one `#[cfg(test)] mod tests` block, and a doc. Production files `gating.rs`, `chat/turn.rs`, `session_channel.rs` confirmed **untouched**; the sole `auto_mode_exec.rs` hunk (`@@ -360,4 +360,80 @@ mod tests {`) is inside the pre-existing test module (excluded from non-test builds).

## Step 16b: Outside-In Testing Results

Re-validated the branch from the consumer (`cargo`) boundary as an external user of these test suites. **0 fixes required** — the characterization suite was already green.

**Detected toolchain**
- Language: Rust (workspace `Cargo.toml`, resolver 2). Toolchain: `cargo 1.97.0` / `rustc 1.97.0`.
- Changed files (`origin/main...HEAD`): `crates/amplihack-signal/tests/chat_it.rs`, `crates/amplihack-signal/tests/session_channel_it.rs` (test binaries), `crates/amplihack-launcher/src/auto_mode_exec.rs` (additions confined to the existing `#[cfg(test)] mod tests`), `docs/testing/TURN_DRIVER_CHARACTERIZATION_TESTS.md` (docs). No production logic touched.
- Package manager / boundary: native `cargo` (Rust CLI repo — no gadugi framework required per qa-team skill).

**Chosen strategy**: run the added characterization tests through the same `cargo` invocations a downstream user/CI would use, under the same feature flags as the existing signal tests, plus the four cargo quality gates. Hermetic, no network/process.

**Scenarios**

| # | Scenario | Command | Result | Key output |
|---|----------|---------|--------|------------|
| 1 (simple) | INV-7 auto-mode stop-condition, default build (no `signal` feature) | `cargo test -p amplihack-launcher characterization_inv7_automode_continues_until_stop_condition` | PASS | `characterization_inv7_automode_continues_until_stop_condition ... ok` — 1 passed |
| 2 (edge/integration) | INV-3/4/5/6 feature-gated signal turn-driver suite | `cargo test -p amplihack-signal --features signal` | PASS | `characterization_inv3_inbound_gate_failclosed ... ok`, `..._inv4_capacity_from_operator_config_evicts_oldest ... ok`, `..._inv5_successive_turns_reuse_same_session_id ... ok`, `..._inv6_turn_runs_to_completion_no_wallclock_cap ... ok` (all suite tests ok) |
| 3 | CLI signal feature integration builds/passes | `cargo test -p amplihack-cli --features signal signal_chat_it` | PASS | all matched binaries ok (no CLI production/test files changed by this PR) |

**Cargo gates (all four green)**

| Gate | Command | Result |
|------|---------|--------|
| Format | `cargo fmt --check` | clean |
| Lint | `cargo clippy -p amplihack-signal -p amplihack-launcher -p amplihack-cli --features amplihack-signal/signal,amplihack-cli/signal --all-targets` | no warnings |
| Build | `cargo build -p amplihack-signal -p amplihack-launcher -p amplihack-cli --features amplihack-signal/signal,amplihack-cli/signal` | Finished |
| Test | scenarios 1-3 above | all passed |

**No-production-change check**: `git diff origin/main...HEAD` limited to two test files, a docs file, and `#[cfg(test)]`-only additions in `auto_mode_exec.rs`; `gating.rs`, `turn.rs`, and `session_channel.rs` are untouched.

**Fix count**: 0 (all scenarios and gates passed on first run at commit 94ee1cfb).

### CODE REVIEW

**Comment by rysweet** [comment]
## Step 17b — Comprehensive Code Review

**Verdict: APPROVE (scoped acceptance gates green).** Test-only PR; zero production logic changed. All five added characterization tests build, pass, and are clippy-clean under `-D warnings`.

### Verification performed locally
| Gate | Result |
| ---- | ------ |
| `cargo fmt --all --check` | ✅ pass |
| `cargo clippy -p amplihack-signal --features signal --tests -- -D warnings` | ✅ pass |
| `cargo clippy -p amplihack-launcher --tests -- -D warnings` | ✅ pass |
| `cargo test -p amplihack-signal --features signal --test chat_it --test session_channel_it` | ✅ 48 + 4 passed, 0 failed |
| `cargo test -p amplihack-launcher --lib auto_mode_exec` | ✅ 8 passed (incl. INV-7) |

### Review checklist
- [x] **Code quality & standards** — tests read cleanly, each carries a rationale header explaining the refactor-critical invariant being locked.
- [x] **Test coverage adequate** — INV-1..INV-7 all mapped in `docs/testing/TURN_DRIVER_CHARACTERIZATION_TESTS.md`, pre-existing vs added clearly distinguished.
- [x] **No TODOs / stubs / swallowed exceptions** — none present in the diff.
- [x] **No unimplemented functions** — n/a (test-only).
- [x] **Logic correctness** — verified below.
- [x] **Edge case handling** — deny-by-default asserted positively (INV-3); FIFO eviction order asserted (INV-4).

### Detailed findings

**INV-7 (`auto_mode_exec.rs`)** — Correctly replays the `while should_continue()` execute loop against the real in-crate `should_continue()` / `build_execution_prompt()` with no process spawn. Marker format `[Turn {turn+1}/{max_turns}]` matches production (`auto_mode_exec.rs:271-273`). Default config's other stop limits (api_calls/output_bytes/session_secs) are large enough that only `max_turns` gates the loop, so the assertion of exactly `MAX_TURNS` prompts is sound. Placement inside the existing `#[cfg(test)] mod tests` keeps it test-only — no non-test build impact. ✔️

**INV-5 (`chat_it.rs::turn`)** — `RecordingRunner` captures argv across two sequential turns and asserts (a) identical pinned `--session-id` continuity and (b) each attacker-shaped prompt survives as exactly one argv element (metacharacters intact, single prompt flag). Strong injection-safety anchor. ✔️

**INV-6 (`chat_it.rs::turn`)** — `GatedRunner` uses tokio oneshot channels for deterministic sync (entry signal + explicit release); proves the turn stays in-flight while blocked and only completes on release. No sleep-as-synchronization. A future loop-level timeout would resolve the turn before release and fail the test — exactly the regression guard intended. ✔️

**INV-3 (`chat_it.rs::gating`)** — Consolidated end-to-end fail-closed anchor over the real `SignalSession::pump_once` + `FakeSignalEndpoint`. Exercises accept + four distinct rejection reasons (wrong-group, wrong-sender, empty-body, echo-within-TTL) and asserts deny-by-default positively via both `accepted` and `drain()`. ✔️

**INV-4 (`session_channel_it.rs`)** — Drives capacity through the operator-config path (`AMPLIHACK_SIGNAL_INBOX_CAPACITY` → `default_capacity()` → `at_session`) rather than an explicit `new(_, N)`, then asserts oldest-evicted FIFO order. Env mutation is save/restored before assertions so a panic cannot leak state. ✔️

### Non-blocking observation (already tracked)
- `session_channel_it.rs` mutates process-global env via `unsafe set_var`, serialized by a file-local `StdMutex`. Since each Rust integration test file is a separate binary, cross-file races aren't possible; within-file serialization is handled. Hardening (serial_test or an `Inbox::with_capacity` injection seam) is legitimately out of scope for a characterization-only PR and is tracked in **#1084**. Not blocking.

**Conclusion:** Meets all scoped acceptance gates. No blocking issues. Approve.

**Comment by rysweet** [comment]
## Step 17c — Security Review

**Verdict: PASS (no security findings).** Test-only PR — 3 test files + 1 doc, **709 additions, 0 deletions, zero production logic changed**. No new attack surface is introduced; several changes are net security-*positive* by pinning existing defensive behavior as regression guards.

### Checklist

| Check | Result | Evidence |
|-------|--------|----------|
| Security requirements met | ✅ | Fail-closed gating, injection-safe argv, and bounded queue are all locked as characterization guards (INV-3/4/5). |
| New vulnerabilities | ✅ None | No production code changed; only `#[cfg(test)]`/`tests/` additions + docs. No new deps. |
| Sensitive data handling | ✅ | No secrets/credentials/tokens in the diff (scanned). Test data uses fabricated numbers (`+1555…`), fake group id, and zero-UUID session ids only. |
| Authn/authz | ✅ | INV-3 test pins **deny-by-default**: empty-allowlist-denies-all, non-allowlisted sender denied, non-primary/own-device sync echoes denied, spoofed-account sync denied, echo-TTL suppression. Any flip to fail-open breaks the test. |
| Injection | ✅ | INV-5 verifies the prompt is passed as **exactly one `-p` argv element**, verbatim, for adversarial inputs (`; rm -rf / #`, `$(whoami)`, `` `id` ``, `&&`, quotes, newlines, leading `-`). |

### Verification of the injection-safety claim (grounded in production)

INV-5 is not a tautological test — it locks a real production boundary:
- `crates/amplihack-signal/src/chat/turn.rs::build_turn_argv` builds a `Vec<String>` with the prompt as a single element.
- `CopilotTurnRunner::run_argv` executes via `Command::new(&program).args(&argv)` — **no `sh -c`, no shell interpolation** — with `stdin` nulled and `kill_on_drop(true)` (no orphaned children, no PID-reuse window).

So shell metacharacters in an inbound Signal message cannot reach a shell. The characterization test freezes this guarantee against future regressions.

### DoS / resource-exhaustion posture

INV-4 locks the bounded inbox (`AMPLIHACK_SIGNAL_INBOX_CAPACITY`, default 32) so a flood of inbound messages evicts the oldest (`PushOutcome::EvictedOldest`) rather than growing unbounded. Invalid/zero/negative/non-numeric values fail safe to `DEFAULT_CAPACITY` (never unbounded, never disabled) — good fail-safe default.

### Non-blocking observation (not a security finding)

`EnvGuard` in `session_channel_it.rs` mutates process-global env via `unsafe std::env::set_var/remove_var`. This is **test-only** and per-test-binary isolated; it is a data-race-UB hardening concern in a multi-threaded harness, **not** an exploitable vulnerability. Already tracked as follow-up #1084. No action required in this PR.

**No blocking security issues. Approved from a security standpoint.**

**Comment by rysweet** [comment]
## Step 17d — Philosophy Guardian Review

**Verdict: PASS.** Test-only PR (709 additions, 0 deletions, 4 files: 3 characterization tests + 1 doc). Zero production logic touched. Fully aligned with amplihack philosophy.

### Philosophy compliance
- [x] **Ruthless simplicity** — Each test replays the *real* in-crate decision path (`AutoMode::should_continue`/`build_execution_prompt`, `Inbox` capacity/eviction, `SerialTurnDriver` argv/session-id) with no process spawn and no superfluous scaffolding. INV-7 replays the actual `while should_continue()` loop rather than re-deriving arithmetic. Minimal means to lock each invariant.
- [x] **Bricks & studs** — Tests exercise stable public seams (the "studs"): `Inbox::at_session`/`default_capacity`/`push`/`drain`, `AutoMode` config+loop, `build_turn_argv`. The doc explicitly names the shared `Session`/`Channel` stud that later PRs will extract, so the contract is defined before the refactor. Each test file stays within its crate boundary.
- [x] **Zero-BS (no stubs / faked APIs / swallowed exceptions)** — No `todo!`/`unimplemented!`/`FIXME`. No faked production APIs — real functions are invoked. Test doubles (channel-gated mock in INV-6, argv recorder in INV-5) are legitimate test seams, not production stubs. `unwrap()`/`expect()` are idiomatic test assertions that fail loudly — nothing is swallowed.
- [x] **No over-engineering** — One test locks exactly one invariant; markers/FIFO/fail-closed each asserted directly. The INV-4 env handling (`Mutex` guard + save/restore) is proportionate to the process-global `set_var` hazard, not gold-plating.
- [x] **Clean module boundaries** — `auto_mode_exec.rs` (launcher), `chat_it.rs` + `session_channel_it.rs` (signal), and the coverage doc are cleanly separated. The doc's INV-1..7 table maps every invariant to its locking test, keeping the safety net discoverable.

### Notes
- INV-4's `unsafe std::env::set_var` is correctly serialized (`ENV_GUARD` `Mutex`) with panic-safe save/restore; the broader per-binary env-isolation hardening is tracked out-of-band in follow-up **#1084** — appropriately deferred for a characterization-only PR that may not introduce production logic.

**No blocking philosophy issues. Approved from a philosophy-compliance standpoint.**
